# **Tarea Nº 3**
Scope: If, Loops, Functions

Curso:          Fundamentos de Programación en Python para Macroeconomía y Finanzas     
Profesor:       Victor Fuentes Campos     

- Fecha y hora de entrega: Miércoles 10 de enero, 11:59am
- Agregue los nombres de las personas con las que discutió esta tarea: ____ Jorge Dongo y Joshua Quispe ______
- Envíe su tarea haciendo el `push` de su código a su repo en GitHub Classroom: 

As of 2022 Q2, there are almost 3.5 million Andriod apps on Google Play and 1.6 million iOS apps on the App Store, and counting. Giving the digital revolution, there is an increase interest in investment opportunities in this sector

<img src="popular_app_stores.png" max-width="300">

Under this context, you will assess a financial committee interested in investing in software development. The goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue. For this purpose, you will access to data from Google Play and App Store.

- Google Play: you can dowload the dataset containing approximately 11k Android apps from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv). It was collected in August 2018
- App Store: you can dowload the dataset containing roughly 7k iOS apps from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv). It was collected in July 2017

## Exploring data

1. Open the two datasets we mentioned above, and save both as lists of lists

In [1]:
# Solution
import pandas as pd


In [2]:
import glob

In [3]:
dataset_list = []
for fname in glob.glob('./.github/*.csv'):
    table = [pd.read_csv(fname)]
    dataset_list.append(table)


In [4]:
dataset_list[0][0]

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7193  1069830936                           【謎解き】ヤミすぎ彼女からのメッセージ    16808960   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   
7196   977965019           みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~    51174400   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD    0.0           2974676               212          3.5   
1         USD    0.0           2161558              1289          4.5   
2         USD    0.0           2130805               579          4.5   
3         USD    0.0           1724546              3842          4.5   
4         USD    0.0           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7192      USD    0.0                 0                 0          0.0   
7193      USD    0.0                 0                 0          0.0   
7194      USD    0.0                 0                 0          0.0   
7195      USD    0.0                 0                 0          0.0   
7196      USD    0.0                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7192              0.0    1.0.9          4+              Games   
7193              0.0      1.2          9+               Book   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   
7196              0.0    1.4.0          4+       Food & Drink   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  
0                  37                1        29        1  
1                  37                0        29        1  
2                  38                5        18        1  
3                  40                5         1        1  
4                  37                4         1        1  
...               ...              ...       ...      ...  
7192               38                5         1        1  
7193               38                0         1        1  
7194               37                2         2        1  
7195               40                0         1        1  
7196               37                0         1        1  

[7197 rows x 16 columns]

2. Print the first few rows of each dataset and find the number of rows and columns of each dataset

In [5]:
# Solution
display(dataset_list[0][0].head(), dataset_list[1][0].head())


#tenemos mejor visualización al imprimirla sin la función print()
#como el output está truncado por su extensión, se recomienda ver como "scrollable element"




id               track_name  size_bytes currency  price  \
0  284882215                 Facebook   389879808      USD    0.0   
1  389801252                Instagram   113954816      USD    0.0   
2  529479190           Clash of Clans   116476928      USD    0.0   
3  420009108               Temple Run    65921024      USD    0.0   
4  284035177  Pandora - Music & Radio   130242560      USD    0.0   

   rating_count_tot  rating_count_ver  user_rating  user_rating_ver      ver  \
0           2974676               212          3.5              3.5     95.0   
1           2161558              1289          4.5              4.0    10.23   
2           2130805               579          4.5              4.5  9.24.12   
3           1724546              3842          4.5              4.0    1.6.2   
4           1126879              3594          4.0              4.5    8.4.1   

  cont_rating        prime_genre  sup_devices.num  ipadSc_urls.num  lang.num  \
0          4+  Social Networking               37                1        29   
1         12+      Photo & Video               37                0        29   
2          9+              Games               38                5        18   
3          9+              Games               40                5         1   
4         12+              Music               37                4         1   

   vpp_lic  
0        1  
1        1  
2        1  
3        1  
4        1

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [6]:
len(dataset_list[0][0])

7197

In [7]:
print("Dimensions of App Store list:", len(dataset_list[0][0]), "rows,", len(dataset_list[0][0].columns), "columns.")
print("Dimensions of Google Play Store list:", len(dataset_list[1][0]), "rows,", len(dataset_list[1][0].columns), "columns.")

Dimensions of App Store list: 7197 rows, 16 columns.
Dimensions of Google Play Store list: 10841 rows, 13 columns.


For more info about the columns in the App Store dataset, visit [this link](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps?select=AppleStore.csv)

## Preparing data

3. As noted before in the [discussion forum](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015), there is an error in the Google Play dataset. Please, remove the entry 10472 and explain shortly why

In [8]:
#Solution

dataset_list[1][0].iloc[10472:10475]

App       Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame            1.9    19.0    3.0M   
10473                  osmino Wi-Fi: free WiFi          TOOLS     4.2  134203   
10474                             Sat-Fi Voice  COMMUNICATION     3.4      37   

         Size     Installs  Type     Price Content Rating             Genres  \
10472  1,000+         Free     0  Everyone            NaN  February 11, 2018   
10473    4.1M  10,000,000+  Free         0       Everyone              Tools   
10474     14M       1,000+  Free         0       Everyone      Communication   

            Last Updated Current Ver Android Ver  
10472             1.0.19  4.0 and up         NaN  
10473     August 7, 2018     6.06.14  4.4 and up  
10474  November 21, 2014     2.2.1.5  2.2 and up

In [9]:
# Lo que encontramos es que para le entrada 10472 no se puso la categoria y  por tanto los demas valores se arrastraron una posicion de derecha hacia la izquierda, con borrar esa entrada basta
dataset_list[1][0].drop(index=10472, inplace=True)

4. Count the number of duplicates in the Google Play dataset. You must build a function to loop through the Google Play dataset

In [10]:
# Solution

def count_dup():
    dict_duplicados = {}
    for element in list(dataset_list[1][0]['App'].values):

        if element not in dict_duplicados.keys():
            dict_duplicados[element] = 1
        elif element in dict_duplicados.keys():
            dict_duplicados[element] +=1

    total_duplicados = 0
     
    for keys, values in dict_duplicados.items():
        if values>1:
            total_duplicados += values-1

    return total_duplicados

print("Number of duplicates in the Google Play dataset:", count_dup())


Numbrer of duplicates in the Google Play dataset: 1181


In [11]:
# podemos validarlo usando duplicated en el mismo dataframe de gogle
dataset_list[1][0]['App'].duplicated().sum()

1181

In [12]:
#####

5. Remove duplicates ...

5. a. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app


In [24]:
list(dataset_list[1][0][['App', 'Reviews']])

'A'

In [26]:
# Solution
app_reviews_dict = {}
for app_reviews_list in list(dataset_list[1][0][['App', 'Reviews']].values):
    aplicacion = app_reviews_list[0]
    resenha = app_reviews_list[1]
 #guarda el valor de cantidad de reviews independientemente de su formato
    if "M" in resenha: #reemplazar la 'M' de millones por su valor numérico
        resenha = int(float(resenha.replace('M', '')) * 1000000)
    else:
        resenha = int(resenha) #de lo contrario, no necesita conversión, se guarda el valor tal cual

    if aplicacion not in app_reviews_dict.keys():
        app_reviews_dict[aplicacion] = resenha
    else:
        if resenha >= app_reviews_dict[aplicacion]:
            app_reviews_dict[aplicacion] = resenha
        else:
            pass



In [29]:
len(app_reviews_dict)

9659

5. b. Use the information stored in the dictionary and create a new dataset, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews).


In [39]:
# Solution
pd.DataFrame.from_dict(app_reviews_dict.items()).rename(columns={0:'aplicacion', 1:'max_review'}).set_index('aplicacion')

max_review
aplicacion                                                    
Photo Editor & Candy Camera & Grid & ScrapBook             159
Coloring book moana                                        974
U Launcher Lite – FREE Live Cool Themes, Hide Apps       87510
Sketch - Draw & Paint                                   215644
Pixel Draw - Number Art Coloring Book                      967
...                                                        ...
Sya9a Maroc - FR                                            38
Fr. Mike Schmitz Audio Teachings                             4
Parkinson Exercices FR                                       3
The SCP Foundation DB fr nn5n                              114
iHoroscope - 2018 Daily Horoscope & Astrology           398307

[9659 rows x 1 columns]

5. c. Explore the cleaned dataset, it should have 9,659 rows

In [40]:
# Solution
print(f"Este dataframe tiene {len(app_reviews_dict)} rows")

Este dataframe tiene 9659 rows


6. Remove non-english apps from both Android and Apple datasets. You should write a function that takes a string, if there's more than three characters that fall outside the ASCII range (0 - 127), then the function should return False (identify the string as non-English), otherwise it should return True. Hint: use the [```ord()```  function](https://www.w3schools.com/python/ref_func_ord.asp)

In [13]:
# For instance
correct_english_name = "Instachat 😜"  # 1 character outside the ASCII range (😜)
incorrect_english_name = "爱奇艺PPS -《欢乐颂2》电视剧热播" # 13 character outside the ASCII range (ie 播)

In [14]:
print(ord("I"))
print(ord("t"))
print(ord("-"))
print(ord("😜"))
print(ord("爱"))

73
116
45
128540
29233


In [50]:
# Solution
apple_dataset = dataset_list[0][0].drop_duplicates(subset='track_name')
google_dataset = dataset_list[1][0].drop_duplicates(subset='App')

In [58]:
def english_apps(valor: str):
    contador = 0
    for element in valor:
        if ord(element) >= 128:
            contador += 1
        elif contador >= 4:
            return False
    return True

In [61]:
english_apps('Mauricio')

M
a
u
r
i
c
i
o


True

In [57]:
apple_dataset.loc[english_apps(apple_dataset["track_name"])]

TypeError: ord() expected a character, but string of length 8 found

7. Loop through each dataset to isolate the free apps in separate lists. Make sure you identify the columns describing the app price correctly. Be careful, prices come up as strings ('0', $0.99, $2.99, etc.), so make sure you're not checking an integer or a float in your conditional statements. How many apps are free?


In [22]:
# Solution
dataset_list[1][0]

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10840 rows x 13 columns]

## Exploring data

 8. Build a frequency table for the ```prime_genre``` column of the App Store dataset, and the ```Genres``` and ```Category``` columns of the Google Play data set. What are the most common genres for each market? You must develop and deploy your function

In [ ]:
# Solution

 9. One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. Based on that exercise, give one app profile recommendation for Google Play.
 
    For the Google Play data set, we can find the required information in the ```Installs``` column. Then, compute the average number of installs by genre. Hint 1: consider only unique values of the ```Category``` column. Hint 2: If an app has, for instance, ```1,000,000+``` assume it as ```1,000,000``` installs. Hint 3: Be careful with a skewed distribution

In [ ]:
# Solution

 10. The App Store dataset doesn´t have the ```installs``` information. Then, you will construct the average rating from the ```rating_count_tot``` column per genre. Based on that exercise, give one app profile recommendation for the App Store

In [ ]:
# Solution